List View: Allow dragging outside the immediate area by passing down a dropZoneElement

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of addressing #43881 and #32438, alternative to #49420 and #50689. Overall, this PR makes dragging to the top and bottom positions of the list view easier.

Allow dragging and dropping within the list view, beyond the bounds of the list view itself, when passed a `dropZoneElement` (an element to be used as the drop zone, typically a parent of the final drop zone ref). Within the list view sidebar, pass the list view container area as a `dropZoneElement` to the `ListView` component.

Then, when the mouse leaves that drop zone area, clear the drop target so that the blue line no longer appears. The result should be that the blue drop indicator line within the list view is now accurate — it will only display when a drop will occur, and it displays across the whole list view area, making it easier to drag to the top and bottom-most positions.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

One of the frustrations with drag and drop in the list view is that it's very difficult to drag to the top and bottom-most positions. This PR hopefully makes that experience feel more flexible and accurate.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Pass an optional `dropZoneElement` element to `useDropZone` and use its element if available, falling back to the one passed down by `useDropZone`.
* Allow this prop at the `ListView` and `useListViewDropZone` levels.
* In the post and site editors (and widget editor), pass a ref of the list view container to the `ListView`.
* Within the list view drop zone logic, when the user leaves the drop zone area, set the target to `null` so that the drop indicator disappears.
* Update nesting logic so that if a user drags below the final item in the list, it won't be treated as a nesting gesture. This helps when a collapsed block is the last item in the list view, as a user dragging all the way to the bottom-most position likely wishes to move the block below rather than nest.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. In the post editor and site editors, experiment with dragging blocks to the top-most and bottom-most positions. Compared with `trunk`, it should be a lot easier to drop to these positions.
2. Test with short list views and really long post content.
3. If the last block in the list is a collapsed container block (e.g. group), then the nesting gesture should only apply when dragging over the bottom half of that block — if you drag further down, the drop indicator should change to be beneath that block rather than inside it.
4. If the last block in the list is an expanded container block with some level of nesting, then dragging to parents should still work when dragging well below the bottom block position.
5. Smoke test that dragging regular blocks in the editor canvas works as on trunk.
6. Smoke test other list view instances (e.g. navigation in editor browse mode) that nothing breaks elsewhere.

## Screenshots or screencast <!-- if applicable -->

Note how in the before version, the Heading looks like it can be dropped to the very bottom of the area by dragging well below, but it cannot. In the after version, the drop is respected correctly.

| Before | After |
| --- | --- |
| ![2023-05-18 17 17 20](https://github.com/WordPress/gutenberg/assets/14988353/766f276b-2cd7-4e66-9b16-2d6ac9ff769d) | ![2023-05-18 17 13 38](https://github.com/WordPress/gutenberg/assets/14988353/6e1d58e4-af37-414a-ad7d-f762cda2fb54) |

Some more details of the behaviour in this PR:

| Dragging up the block hierarchy should still work when cursor is lower than the last block | Nesting a collapsed block only happens if the cursor is over the block (makes it easier to drop below) |
| --- | --- |
| ![2023-05-19 12 00 02](https://github.com/WordPress/gutenberg/assets/14988353/abd06731-3cbb-4321-9809-f37a591022ab) | ![2023-05-19 11 59 31](https://github.com/WordPress/gutenberg/assets/14988353/c390335e-504c-431d-b9a7-a7db0b93e166) |

andrewserong committed:
List View: Allow dragging outside the immediate area by passing down a dropZoneRef from outside the list view

andrewserong committed:
Fix nesting logic so it will only nest if hovering over the bottom half of the item

andrewserong committed:
Update widget sidebar, too

andrewserong committed:
Update comments

andrewserong commented on packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:
```diff
@@ -378,7 +385,12 @@ export default function useListViewDropZone() {
-			( event, currentTarget ) => {
+			( event, currentTarget, action ) => {
+				if ( action === 'clear' ) {
+					setTarget( null );
```
This change ensures that when a user drags outside the drop zone area, the drop indicator line disappears. This resolves a point of confusion where a line was visible, a user releases the mouse button, and is then confused as to why nothing happened.

kevin940726 replied:
Why don't we just call `setTarget(null)` directly in the `onDragLeave` event? That is, don't combine it in this throttled event.

andrewserong replied:
I tried that originally, but it wound up behaving inconsistently — sometimes it would switch it to `null` but then it seemed that the throttled event from dragging over might fire slightly after, and then reset the `setTarget`. I'm not 100% sure what order things were happening in, but moving to the throttled event resolved it.

Happy to re-test it, though, in case I missed something 🙂

andrewserong replied:
Yep, looks like if we have `setTarget( null )` directly in `onDragLeave` then the throttled event fires directly after, restoring the target. It's pretty subtle, but you should hopefully be able to see the flicker off when dragging out of the list view in this gif, before it gets displayed again:

![2023-05-19 14 49 15](https://github.com/WordPress/gutenberg/assets/14988353/81f6f286-3907-4672-a4b5-9950ec635c23)

kevin940726 replied:
I think that's because we're not canceling the throttled event? Try calling `throttled.cancel()` before setting the target and see if it fixes the issue? Nvm if it's not the case 😆 .

andrewserong replied:
Ah, of course! That's working nicely, and I see now that's how the `onDragEnd` callback does it, too. Committed in https://github.com/WordPress/gutenberg/pull/50726/commits/868c57ee766149501c42e193402045b0cce94aa0. Thanks @kevin940726, much appreciated! 🙇

kevin940726 replied:
Nice! Thanks for double-checking! 🙇

andrewserong committed:
Update comments and types

andrewserong committed:
Use setTarget directly in the callback

jasmussen commented:
From the GIFs this feels like a substantial improvement, and something that I've wrestled with countless times. In testing I can't get it to work, though. Testing the post editor:

![branch](https://github.com/WordPress/gutenberg/assets/1204802/c5947cad-d33f-4f4d-bd7b-045c12c1ab90)

What am I doing wrong?

apeatling reviewed (commented):
Confirming what @jasmussen sees. I can't seem to trigger this change either when dragging past the last item in the list. Everything built and working correctly from what I can see?

![2023-05-19 09 54 33](https://github.com/WordPress/gutenberg/assets/1464705/494ae095-52e3-42a8-820c-b6fc0fcf72a6)

andrewserong commented:
Thanks folks! Sounds like I might have broken something in the latests commits, I'll dig into it 👍

andrewserong committed:
Fix issue where sometimes the expanded drop zone area was not being respected

andrewserong commented:
Thanks again for testing and flagging the issue @jasmussen and @apeatling — it turns out I was testing with the `Always open list view` preference running, and so I didn't notice that I missed updating the dependency array in `useDropZone` so that the logic would refresh when the `ref` for the drop zone area changes (which occurs when opening / closing the list view).

I've pushed a fix in https://github.com/WordPress/gutenberg/pull/50726/commits/970217ea32b59f6b4e30c95c53b83c7e77ceec75, so it should be working now, but do let me know if you still run into any issues!

Updated: I've added a couple of simple tests in https://github.com/WordPress/gutenberg/pull/50726/commits/84ac58311087bd65461213532b29b1322fe2ac47 to test that the drop zone is flagged correctly.

andrewserong committed:
Add tests for useDropZone

jasmussen reviewed (approved):
Just delightful:

![nice](https://github.com/WordPress/gutenberg/assets/1204802/de7d4378-22c3-4ddc-b45d-075baa794247)

Probably get a confidence boost on the code, but this feels vastly improved. Thank you!

talldan commented on packages/compose/src/hooks/use-drop-zone/index.js:
```diff
@@ -61,11 +63,16 @@ export default function useDropZone( {
+			// If a custom dropZoneRef is passed, use that instead of the element.
+			// This allows the dropzone to cover an expanded area, rather than
+			// be restricted to the area of the ref returned by this hook.
+			const element = dropZoneRef?.current ?? elem;
```
I'm wondering how this will work in terms of the cleanup of the hook.

Because it's a ref effect, I think the callback will typically only trigger when `elem` is unmounted, not when `dropZoneRef.current` is unmounted.

andrewserong replied:
Good question! I think this is working pretty well so far. To confirm I logged out `element` within the callback, and it appears to be the right element:

<img width="445" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/dede0dbd-0816-4005-b76d-3afba45e756e">

As far as I can tell the behaviour is something like the following:

* We pass in a ref to `useDropZone`, but `useDropZone` also returns a `ref`. That `ref` is still used in the list view, to attach to the `TreeGrid` component.
* When `TreeGrid` unmounts, that results in the cleanup callback being called.
* Because the cleanup callback was instantiated with a reference to `element` being the one passed in by `dropZoneRef` the cleanup callback cleans up the passed in element.
* Also, because we have `dropZoneRef?.current` in the dependencies array of `useRefEffect` when the passed in ref changes, the cleanup appears to correctly fire, too. From checking the docs, this appears to be consistent with [how useEffect works](https://react.dev/reference/react/useEffect#my-cleanup-logic-runs-even-though-my-component-didnt-unmount) in that the cleanup fires not only when the component unmounts, but whenever any of the dependencies change. The doc comment for `useRefEffect` appears to confirm this here, too: https://github.com/WordPress/gutenberg/blob/62b3b0e5fc4f79f3021ea416d90edafb52724f4e/packages/compose/src/hooks/use-ref-effect/index.ts#L11-L29

So, all up, I don't think there are any stray event listeners not being removed, as far as I can tell.

Do let me know if you think I've missed anything here, though!

talldan replied:
> When TreeGrid unmounts, that results in the cleanup callback being called.

Sorry if I'm making this more complicated. Given this is being introduced as a generic stable API, I'm mainly thinking that there could be different ways it's used (or misused):
- The `dropZoneRef` could reference a completely separate part of the DOM that isn't mounted/unmounted at the same time as the effect's component.
- The `dropZoneRef` could reference a parent of the effect's component that isn't mounted/unmounted at the same time.

I don't know if there are tests that can be added to assert these different cases work as expected.

It also might be worth expanding upon the docs for the prop to ensure that the right usage of this API is promoted. There's an example of this for Popover's `anchor` prop that recommends particular usage - https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/popover/README.md#anchor-element--virtualelement--null.

> Also, because we have dropZoneRef?.current in the dependencies array of useRefEffect when the passed in ref changes, the cleanup appears to correctly fire, too

This is the part that often trips me up. With `dropZoneRef` being a ref type, it won't trigger a re-render when it changes. I've sometimes opted to store an element using `useState` for that reason (which is also what `Popover` recommends).

andrewserong replied:
> Sorry if I'm making this more complicated.

No apology necessary! I'm happy to go as slowly as needed with this PR — I always like to be a bit more cautious when making changes to lower-level packages such as `compose`, and it's important we get it right and don't introduce any issues 👍

> I don't know if there are tests that can be added to assert these different cases work as expected.

Good example use cases, it's worth digging in, I'll do a little more exploration to see if it's possible to come up with additional test cases. I think expanding the test cases there is probably a good path toward building confidence around the API change.

> It also might be worth expanding upon the docs for the prop to ensure that the right usage of this API is promoted. There's an example of this for Popover's anchor prop that recommends particular usage

Good idea, for some reason `useDropzone` isn't documented in the `compose` package's readme, but I'll have a go at writing an entry for it.

> This is the part that often trips me up. With dropZoneRef being a ref type, it won't trigger a re-render when it changes. I've sometimes opted to store an element using useState for that reason (which is also what Popover recommends).

Ah, good point, because the ref reference itself isn't changing, only `current` 🤔. In practice, I haven't yet run into any issues in this PR so far. In terms of using `useState`, do you mean we'd do that internally within `useDropZone`, or are you imagining that the consumer would use `useState`, and we'd pass the element directly to `useDropZone` instead of a `ref`?

Edit: I think you probably mean the latter — and I can see that the url popover for example, uses the state approach here: https://github.com/WordPress/gutenberg/blob/493eac0d2a1946eb0543f56830dfcf347b6eadd3/packages/block-editor/src/components/url-popover/image-url-input-ui.js#L251 — happy to swap it over to that pattern here 👍

talldan replied:
> Edit: I think you probably mean the latter — and I can see that the url popover for example, uses the state approach here

Yep, exactly that 👍

andrewserong replied:
Update: I've done the following:

* Swapped out `dropZoneRef` for a `dropZoneElement` and updated usage to use state for storing the element before passing it along to the `ListView` (and then on to `useDropZone`)
* Added a README file for `useDropZone` — because the hook is still currently exported as experimental, I couldn't add directly to the `compose` package readme. I see a couple of other experimental hooks also include their own README files in the `compose` package, so this seemed like a decent place to store the documentation for now, while the hook is still "experimental". Happy for any feedback there, I found it a bit challenging coming up with straightforward wording 😅 

I haven't looked into exploring adding other tests just yet, but will dig in tomorrow.

Thanks again for the feedback!

talldan replied:
> because the hook is still currently exported as experimental, I couldn't add directly to the compose package readme.

Thanks, didn't realise it was experimental. It must have been that way for ages!

My hunch is that using state to store the element will probably solve the issues, hopefully that's the case.

talldan commented on packages/compose/src/hooks/use-drop-zone/index.js:
```diff
@@ -228,6 +235,6 @@ export default function useDropZone( {
 			};
 		},
-		[ isDisabled ]
+		[ isDisabled, dropZoneRef, dropZoneRef?.current ] // Refresh when the passed in dropZoneRef changes.
```
There's possibly only a need for `dropZoneRef?.current` here?

I think the ref object by definition shouldn't change.

andrewserong replied:
Good catch, yes, we only need to look for `dropZoneRef?.current` 👍
Updated in https://github.com/WordPress/gutenberg/pull/50726/commits/4eb834dfffa4bb2cf1d301ffb3095a585ded882b

kevin940726 replied:
IIRC, putting `ref.current` inside the dependency array should trigger lint warning. Since updating `ref.current` won't trigger a re-render, it doesn't make sense to put that in the dependency array.

andrewserong replied:
Thanks 👍
I've pushed a few changes, so we're now passing a `dropZoneElement` param instead of a `ref`, so the dependency array now just has `isDisabled` and `dropZoneElement` in it.

kevin940726 replied:
Here are some related articles about this:

- https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array/
- https://stackoverflow.com/a/60476525

kevin940726 replied:
Oh nice! I didn't noticed that. Ignore this please 😆 .

andrewserong replied:
All good — I appreciate the links and extra knowledge as always! 🤗

talldan reviewed (commented):
This is working well. I'm unsure of some the specifics around how the `useDropZone` hook works, I personally always find these things a little hard to understand though!

andrewserong committed:
Update dependency array, we only need to look at the current key

andrewserong committed:
Swap out dropZoneRef for dropZoneElement

andrewserong committed:
Fix typo

andrewserong committed:
Add README file

apeatling reviewed (approved):
Working great now, a big usability improvement!

![2023-05-23 11 31 16](https://github.com/WordPress/gutenberg/assets/1464705/39020603-9742-478d-a238-484fb0e63bf7)

andrewserong commented:
Thanks for the re-reviews!

@talldan and @kevin940726 — unfortunately I haven't had much luck writing additional tests for this, as I can't seem to get the event handlers to fire within an individual test (I've tried a `user.pointer` approach, a `fireEvent.mouseDown` / `fireEvent.mouseMove` approach, and firing `dispatchEvent` directly on the expected elements).

At this stage, do you think adding more tests is a blocker to landing this PR? Happy to keep looking into it, but I'm currently a bit stumped as to how to reliably write tests for this hook where the drag events are fired 🤔

talldan commented:
> At this stage, do you think adding more tests is a blocker to landing this PR?

I don't really know a lot about the testing situation, so I'm not very useful in offering suggestions!

I think it's good to merge given the docs now recommend using the hook in the same way List View uses it (and that's working well).

talldan commented on packages/compose/src/hooks/use-drop-zone/README.md:
```diff
@@ -0,0 +1,71 @@
+# useDropZone
```
```suggestion
# useDropZone (experimental)
```

Until this is stabilized, I think the docs should make clear it's experimental.

Potentially already a few plugins using it though when looking at WPDirectory.

andrewserong replied:
Good idea, thanks!

andrewserong committed:
Update packages/compose/src/hooks/use-drop-zone/README.md

Co-authored-by: Daniel Richards <daniel.richards@automattic.com>

andrewserong commented:
> I think it's good to merge given the docs now recommend using the hook in the same way List View uses it (and that's working well).

Thanks Dan! I've just committed your Readme suggestion, and once the GH actions pass, I'll merge this in.
